### ISSUE TITLE
What do "internal" and "external" mean in value flows?

### ISSUE THREAD

**Issue by bhaugen** [valueflows/valueflows#133]
See https://github.com/valueflows/process/issues/18#issue-157317535

@elf-pavlik wrote:
> It looks like we can distinguish 6 sources for allocating resources as inputs for the process:
'internal'
[...]
'external'
[...]
The 'internal' and 'external' seem to differ only that it may require a vf:Transfer, while internal we can do allocation without any transfers. 

I replied:
> 'Internal' and 'external' are interesting concepts that deserve their own thread. Internal to what? (That's why we have contexts, or context agents.) What does 'someone else's inventory' mean? What does it mean in a mutual coordination economy?

This is that thread (to discuss internal and external to what).

**Comment by elf-pavlik** [comment]
In [Shares: Peanuts deposited in community storage](https://github.com/valueflows/valueflows/blob/master/use-cases/shares.md#peanuts-deposited-in-community-storage) case we also find:

> - People can transfer those shares to each
> - When someone requests donation of peanuts, shareholders can meet that requests by combining the total weight from any combination of fractions of their shares

I could add, that community kitchen could also request more peanuts and people could donate some of their shares. In case above each person acts as member of community but also as independent 'external' agent. As well as each of those people can have membership in _n_ communities.

**Comment by elf-pavlik** [comment]
It looks like in NRP you model differently _work_ and _use_ events when involved agents are all members of the 'context' agent and when they don't have this member relationship.
- work rea:EconomicEvent vs. service rea:EconomicResource
- use rea:EconomicEvent vs. rental service rea:EconomicResource

I would like to explore possibility that we can have uniform way of modelling work and usage which doesn't change for different relationships between agents. And where that work and usage can also always appear in context of intent and transfer.

**Comment by bhaugen** [comment]
Work in NRP is always modeled the same way. Likewise usage. Both are rea:EconomicEvents.

**Comment by fosterlynn** [comment]
@elf-pavlik I generally appreciate where you are going with taking independent views across networks.  But continuing to explore:

Part of this discussion might be how people think of their networks.  And how to cross boundaries of networks or other groups or organizations.  (NRP of course has technical boundaries which we should ignore, since it has an instance supporting one or a related set of networks or groups.)

In Sensorica, they would think of anyone they know, doing work that they can see, on one of their projects, as part of the Sensorica network. You are a member (they call it "affiliate").  They are more formal than that, that person would also have an agent relationship with a project or with Sensorica.  They also have an understanding of what is the Sensorica commons in terms of equipment that gets used in projects, including people's personal equipment brought in, sometimes with access rules, etc. So the context agent becomes important - not talking about software, but about how people tend to think of the groups they work with.

In terms of an external entity that is part of the old economy, say "renting" a piece of equipment from the local hardware store:  The piece of equipment never enters Sensorica inventory, they basically just pay for the rental service (you could call it use, but you don't want to know everything about that equipment that you would need to in order to vf:use it). Same would be if they hired someone to clean the lab (a funny example of course), they don't know anything about that person, to them it is only a service.

Another note: The way the value equation works, any project can set out their own value equation. So if Project A uses something from Project B's equipment commons, the distribution process would just send the required transfer amount to Project B, so that they can run their own value equation.  At that point it is something like a service to Project A, and it is Project B's right to distribute from there.  (This might not exactly fit with NRP's code, it definitely does for consumption of components; not sure for equipment, but if it doesn't, it probably should.)

There are a lot of use cases in between doing it yourself and hiring someone or renting something from the old economy, of course.  Will some time soon try to set out some use cases.

**Comment by elf-pavlik** [comment]
Thanks @fosterlynn 

I think writing down YAML snippets for _vf:Request_ and _vf:Transfer_ of work and usage can help with clarifying it all further. IMO any kind of input to the process needs to work very well with _vf:Request_ and _vf:Transfer_. Also any kind of output needs to work very well with _vf:Offer_ and _vf:Transfer_. Requesting/offering and transferring build around instances of  vf:Resource currently seems more intuitive for me. At the same time I stay open to look at examples of _vf:Intent_ and _vf:Transfer_ which work with **vf:Process** related events. As I understand your 'exchange' oriented diagrams, they connect instances of _vf:Transfer_ to instances of _vf:Resource_ and not directly to instances of _vf:Process_ related events or commitments.

**Comment by bhaugen** [comment]
> Requesting/offering and transferring build around instances of vf:Resource currently seems more intuitive for me. 

More intuitive than what? And are you sure your intuition is a good guide?

**Comment by fosterlynn** [comment]
> As I understand your 'exchange' oriented diagrams, they connect instances of vf:Transfer to instances of vf:Resource and not directly to instances of vf:Process related events or commitments.

Yes.  

> Requesting/offering and transferring build around instances of vf:Resource currently seems more intuitive for me.

I think we agree on that.  (Assuming resource / resource type / etc, not getting hung up on actual touchable resource.)

> At the same time I stay open to look at examples of vf:Intent and vf:Transfer which work with vf:Process related events. 

Why?  How would that work?

**Comment by elf-pavlik** [comment]
> > At the same time I stay open to look at examples of vf:Intent and vf:Transfer which work with vf:Process related events.
> 
> Why? How would that work?

As I understood, you prefer to model 'work' and 'usage' as 'process related' rea:EconomicEvent (an input) and not rea:EconomicResource. In such case we need a way to cast intents (make requests) for those process input events, and also build transfer around those process input events (not the resources).

We could analyse it on two examples:
- use a car for delivering food (blocking it for 6 hours on particular day)
- drive a car for delivering food (occupying oneself for 6 hours on particular day)

Agents
- Food Not Bombs (CDMX) - organizes the food and coordinates the delivery
- Alice - owns the car
- Bob - drives the car

All agents stay independent, Alice and Bob don't have _member_ relationship with FnB. Alice transfers usage of her car as donation to FnB and Bob transfers his work as donation to FnB. FnB published two requests, and this way found Alice and Bob. We can also add variant with exchanges, where FnB does reciprocal transfers to both Bob and Alice for 2kg of peanuts to each one.

**Comment by bhaugen** [comment]
> As I understood, you prefer to model 'work' and 'usage' as 'process related' rea:EconomicEvent (an input) and not rea:EconomicResource. 

Do you understand why? This is an ongoing circular argument. We should find a way to come to a firm conclusion.

**Comment by fosterlynn** [comment]
I see better where you are coming from.  

> We could analyse it on two examples:
>     use a car for delivering food (blocking it for 6 hours on particular day)
>     drive a car for delivering food (occupying oneself for 6 hours on particular day)

So, thinking a bit about Intents (which I have not done yet very deeply), I think we could model it something like a rea:Commitment, and would _reference_ a  rea:ResourceType (probably more than an actual rea:EconomicResouce, although it could perhaps be that too).  I think intents for rea:EconomicEvents are very messy.  But I also think this is mostly a modeling question, which of course is what is most useful.  (Map not the territory.)

I'll take some time to suggest a model that includes everything we have talked about on the "plan" middle layer, and uses rea as a base.  I'll use this use case, although there is also the more general offer and want request that is not dependent on a specific future process and timing.  Will try for today, but not sure. 

Although perhaps this gets away from "internal" and "external"?  I'll try to do it assuming Alice and Bob are somehow "external", just random donations to FnB.

**Comment by elf-pavlik** [comment]
>  I think intents for rea:EconomicEvents are very messy.

I have similar impression, also thinking about transfers of those rea:EconomicEvents (process inputs and outputs) gives me the same impression. As I understand casting intents leads to transfers...

> I'll take some time to suggest a model that includes everything we have talked about on the "plan" middle layer, and uses rea as a base. I'll use this use case, although there is also the more general offer and want request that is not dependent on a specific future process and timing.

During our last call, we discussed focusing on [Planning with the vf:Vocabulary](https://github.com/valueflows/valueflows/wiki/Planning-with-the-vf:Vocabulary) which leads to [The evolution of intents to commitments to claims to economic events](https://github.com/valueflows/exchange/issues/23).
I think this use case fits it very well
- FnB plans a process with particular dates
- FnB casts intents _vf:Request_ which link to that planned process or even to relevant planned inputs
- Alice and Bob respond to those requests with their proposals of donating usage and work
- FnB receives transfers of usage and work that go into that process ([in-kind donations](https://en.wikipedia.org/wiki/Gifts_in_kind))

> Although perhaps this gets away from "internal" and "external"? I'll try to do it assuming Alice and Bob are somehow "external", just random donations to FnB.

:+1:
Once we have it donation based, we can always add reciprocal transfers for peanuts, cookies or whatever
- FnB makes reciprocal transfers to Alice and Bob for 2kg of peanuts for each one (reciprocal but also voluntary in-kind donation, Alice and Bob didn't expect receiving anything back and had no claims for it)

**Comment by elf-pavlik** [comment]
Conversation above seems very relevant to [Context agents](https://github.com/valueflows/valueflows/pull/149) PR.

> Work in NRP is always modeled the same way. Likewise usage. Both are rea:EconomicEvents.

We have recently added to `valueflows/resource` _vf:WorkResource_ and _vf:UsageResource_. I have impression that NRP focuses mostly as using work and usage on inputs to processes. At the same time not providing way to model processes that **output** work and usage (the 'time based stock'- availability?). If we model work and usage just as events, we have no resource that would appear as input of one process and output of another process (IPO), attempting to describe process that creates work and usage may result in need to have event that 'connects' two processes and has nature of input and output depending on which process we look at it. The resource seems to me as more fitting construct to connect processes and have input & output **relationship** to them.

Conversation in comments above also touches casting intents for work and usage and making transfers of work and usage. I wonder if you assume that casting intents only happens _between context_ and one would not cast intents _within single context_ ?

**Comment by bhaugen** [comment]
> At the same time not providing way to model processes that output work and usage (the 'time based stock'- availability?). 

I'll respond about inanimate usage resources first. Human labor is a lot more complicated.

Inanimate tangible resources that are used in time increments are created like any other tangible resources, as outputs from manufacturing or construction processes, usually measured in Eaches (units of count, not units of use).

[Edit: or they maybe gotten via exchanges, like the recurrent 3d printer example.]

**Comment by elf-pavlik** [comment]
> Inanimate tangible resources that are used in time increments are created like any other tangible resources, as outputs from manufacturing or construction processes, usually measured in Eaches (units of count, not units of use).

What about the maintenance? Sometimes 'custodian' also needs to contribute some work to make some tool available...

We may need some use cases which capture [Lease](https://en.wikipedia.org/wiki/Lease)

**Comment by bhaugen** [comment]
> What about the maintenance? Sometimes 'custodian' also needs to contribute some work to make some tool available...

Absolutely! Those are the lovely processes where the same resource is an input and an output. The tool is still an "Each". But the usage processes have reciprocal responsibilties for the tool and its original acquisition or creation as well as its ongoing maintenance.

**Comment by bhaugen** [comment]
Another angle on internal and external and contexts is who wants to know what details about which events. 

For example, in transportation, we might have the source context, the transportation context, and the destination context.

The source context is probably interested in the effects on their inventory, who issued and packed the goods for shipping, and who loaded the goods on the truck (for example). The transportation context wants to know when and where they need to pick up what, the manifest of the truck: what's on it, and where it needs to go. They might or might not want to know what's in the package, depending on whether the contents are dangerous, alive, or some other aspects that might affect them in some way. The destination context wants to know what got unloaded, who unloaded and unpacked it, what they received, and where they put it. Both the source and the destination want to know the FOB point and all contexts want to know who is responsible in cases of damages, and who is responsible for paying for what when, especially if they need to pay or can expect payment.

[Edit: I realize I am pesonifying contexts here, while who really wants to know what is some person who is responsible for knowing that information in their context.]

**Comment by elf-pavlik** [comment]
> and who is responsible for paying for what when, especially if they need to pay or can expect payment.

Rephrasing: "and who agreed to give what and when in exchange, especially if they need to give something in exchange or can expect receiving something in exchange."

Even if we would agree among ourselves to interpret "pay" as "give something in exchange", while communicating this way to people outside of this agreement, they might struggle not to fail into interpreting "pay" as "give $ in exchange" instead of leaving the 'resource type' as variable "give ? in exchange".

**Comment by bhaugen** [comment]
I agree with your rephrasing for value flows. I was just depicting typical commercial transportation relationships.

**Comment by fosterlynn** [comment]
> Conversation in comments above also touches casting intents for work and usage and making transfers of work and usage. I wonder if you assume that casting intents only happens between context and one would not cast intents within single context ?

An example where intents are cast within a context all the time is time banks.  And probably lots of marketplace kind of associations, like food networks, coops of exchangers, etc.
